pkg/controller/source/source.go

```diff
@@ -46,18 +52,113 @@ type Source interface {
 	Start(eventhandler.EventHandler, workqueue.RateLimitingInterface, ...predicate.Predicate) error
 }
 
+var _ Source = &ChannelSource{}
+
 // ChannelSource is used to provide a source of events originating outside the cluster
 // (eh.g. GitHub Webhook callback).  ChannelSource requires the user to wire the external
 // source (eh.g. http handler) to write GenericEvents to the underlying channel.
-type ChannelSource chan event.GenericEvent
+type ChannelSource struct {
+	// once ensures the event distribution goroutine will be performed only once
+	once sync.Once
+	// source is the source channel to fetch GenericEvents
+	source chan event.GenericEvent
```

Comment:
Wondering if this should be a read channel only.  What do you think?

Reply:
good catch